Treat -Target as literal in New-Item

GameMicrowave opened the pull request:
# PR Summary

<!-- Summarize your PR between here and the checklist. -->

Fix #6232
Fix #13136

Fix `New-Item` fails to create file links or creates links pointing to incorrect locations when the target path contains wildcard characters.

Make the `-Target` and `-Value` parameters of `New-Item` treat the path as a literal.

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

When using the `New-Item` command to create file links, if the target path contains wildcard characters, they are incorrectly interpreted. The following scenarios occur:

+ When creating a hard link, if the link path is an absolute path, the creation will always fail.
+ When creating a symbolic link:
  + If the wildcard characters in the link path are not escaped with a backtick, the creation will fail.
  + If the wildcard characters in the link path are escaped with a backtick, a link pointing to an incorrect path will be created.
<!--  + if the specified path is not the current directory and the target path is a relative path, the link will point to the wrong location. This is because relative paths are interpreted based on the current working directory at the time the link is created, not the directory where the link resides. -->
+ When creating a Junction, the situation is similar to the above.

Example:

```powershell
New-Item lk21.txt -ItemType HardLink -Target 'D:\[test].txt'
# New-Item: Could not find item D:\[test].txt.

# The `Get-Item` command can correctly retrieve the file.
Get-Item -LP 'D:\[test].txt'
#    Directory: D:\
# Mode                 LastWriteTime         Length Name
# ----                 -------------         ------ ----
# -a---           2025/3/15    14:26              9 [test].txt
```

Now, the path is correctly parsed to create the link.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
  - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
  - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
  - [ ] None
  - **OR**
  - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.5/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
    - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
  - [x] Not Applicable
  - **OR**
  - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
  - [x] N/A or can only be tested interactively
  - **OR**
  - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
  - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
  - **OR**
  - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
    - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
      - [ ] Issue filed: <!-- Number/link of that issue here -->
    - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
      - [ ] Issue filed: <!-- Number/link of that issue here -->

GameMicrowave committed:
Update SessionStateContainer.cs

Fix Path explain error

GameMicrowave committed:
Update regex.cs

Fixed char '`' can not escape

GameMicrowave committed:
Update FileSystemProvider.cs

Fixed wildcard conversion of hard links

GameMicrowave committed:
Update FileSystemProvider.cs

Fix wildcard conversion of hard links

GameMicrowave commented:
> 请阅读以下贡献者许可协议 （CLA）。如果您同意 CLA，请回复以下信息。
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > 选项：
> > 
> > * （默认 - 未指定公司）我对我的提交内容拥有唯一的知识产权，并且我不会在为我的雇主工作期间提交提交内容。
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * （当给出公司时）我在为我的雇主工作期间提交了提交内容（或者我的雇主根据合同或适用法律对我的提交内容拥有知识产权）。我已获得雇主的许可，可以代表我的雇主提交内容并签订本协议。通过在下面签名，定义的术语“您”包括我和我的雇主。
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> 贡献者许可协议

@microsoft-github-policy-service agree

iSazonov commented:
@GameMicrowave There is #16306 which must be fixed first since it is public API.

ArmaanMcleod commented on src/System.Management.Automation/engine/regex.cs:
```diff
@@ -240,7 +240,7 @@ internal static string Escape(string pattern, char[] charsNotToEscape)
                 // if it is a wildcard char, escape it
                 //
-                if (IsWildcardChar(ch) && !charsNotToEscape.Contains(ch))
+                if ((IsWildcardChar(ch) || ch == '`') && !charsNotToEscape.Contains(ch))
```
```suggestion
                if ((IsWildcardChar(ch) || ch == escapeChar) && !charsNotToEscape.Contains(ch))
```

ArmaanMcleod replied:
@iSazonov I can't remember if this was the actual correct fix to the `Escape` method. We probably need to have a 2nd attempt at fixing public API for this in another PR.

iSazonov replied:
Let's discuss in related issue.

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave committed:
Update regex.cs

replace literal char with field

GameMicrowave committed:
Merge branch 'master' into master

iSazonov commented:
@GameMicrowave Please resolve merge conflicts.

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave commented:
> @GameMicrowave Please resolve merge conflicts.

yes, i resolved it😊

iSazonov commented:
@GameMicrowave I tried the build with examples from related issues and they don't work. Please add tests if you want to continue working on this issue.

GameMicrowave committed:
Update FileSystemProvider.cs

fix relative path issue

GameMicrowave committed:
Update New-Item.Tests.ps1

Add Test of file link creation

GameMicrowave commented:
> @GameMicrowave I tried the build with examples from related issues and they don't work. Please add tests if you want to continue working on this issue.

Alright, I've written a test script. In the process, I discovered another bug, which I've now fixed.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1:
```diff
@@ -396,3 +396,34 @@ Describe "New-Item -Force should throw an error for invalid characters in direct
+
+        New-Item $hardLink -Target ([WildcardPattern]::Escape((Get-Item -PSPath $testFile))) -Type HardLink    # create hardlink with absolute path
+        New-Item $symbolicLink -Target ([WildcardPattern]::Escape($testFile)) -Type SymbolicLink
+        New-Item $junction -Target ([WildcardPattern]::Escape((Get-Item -PSPath $testFolder))) -Type Junction
```
Please see PowerShell Committee conclusion in https://github.com/PowerShell/PowerShell/issues/6232#issuecomment-1864855190

GameMicrowave replied:
OK, I make the `-Target` and `-Value` parameters of `New-Item` treat the path as a literal.

GameMicrowave committed:
Update New-Item.Tests.ps1

make the `-target` treat the path as a literal

GameMicrowave committed:
Update SessionStateContainer.cs

make the `-target` treat the path as a literal

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave committed:
Update New-Item.Tests.ps1

..

GameMicrowave committed:
Update New-Item.Tests.ps1

improve test

GameMicrowave committed:
Update FileSystemProvider.cs

Undo changes

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave committed:
Update New-Item.Tests.ps1

improve test

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave committed:
Update GroupPolicySettings.Tests.ps1

-Tag -> -Tags

GameMicrowave committed:
Update GroupPolicySettings.Tests.ps1

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave commented:
I don't know how to add a CI label to a PR. I tried to find a button, but I didn't find one

GameMicrowave committed:
Update New-Item.Tests.ps1

improve test

GameMicrowave committed:
Merge branch 'master' into master

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1:
```diff
@@ -396,3 +396,42 @@ Describe "New-Item -Force should throw an error for invalid characters in direct
+    }
+    
+	AfterAll {
+        Pop-Location
+        Pop-Location
```
I it is not reliable. I suggest to use full paths as in other tests.

GameMicrowave committed:
Update New-Item.Tests.ps1

just use single pop-location

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1:
```diff
@@ -396,3 +396,40 @@ Describe "New-Item -Force should throw an error for invalid characters in direct
+        $expectedFileContent = 'succeed!'
+
+        New-Item "TestDrive:\$testFolder" -Type Directory
+        Push-Location -PSPath "TestDrive:\$testFolder"
```
Please don't use alias PSPath. We should explicitly use `Path` or `LiteralPath`.

GameMicrowave committed:
Update New-Item.Tests.ps1

use original parameter name

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave commented:
@iSazonov All the tests have been passed. Do you think there's anything else that needs to be modified?

mklement0 commented on src/System.Management.Automation/namespaces/FileSystemProvider.cs:
```diff
@@ -2272,7 +2272,7 @@ protected override void NewItem(
                             if (!IsAbsolutePath(strTargetPath))
                             {
                                 // there is already a check before here so that strTargetPath should only resolve to 1 path
```
This comment can probably be removed.

GameMicrowave replied:
OK

mklement0 commented:
@iSazonov, other than that, it LGTM. Thanks your tackling this, @GameMicrowave.

GameMicrowave committed:
Update FileSystemProvider.cs

remove comment

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave committed:
Merge branch 'master' into master

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/New-Item.Tests.ps1:
```diff
@@ -396,3 +396,40 @@ Describe "New-Item -Force should throw an error for invalid characters in direct
     }
 }
+
+Describe "If the target path contains wildcard characters, the operation should succeed in creating a file link." -Tags @("CI", 'RequireAdminOnWindows') {
```
All file link tests are in FileSystem.Tests.ps1 line 555. Please move new tests to the file.

GameMicrowave committed:
Update New-Item.Tests.ps1

MOVE TEST

GameMicrowave committed:
Update FileSystem.Tests.ps1

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/FileSystem.Tests.ps1:
```diff
@@ -611,12 +611,12 @@ Describe "Hard link and symbolic link tests" -Tags "CI", "RequireAdminOnWindows"
-        $symLinkToNothing = Join-Path $TestPath "sym-link-to-nowhere"
-        $dirSymLinkToDir = Join-Path $TestPath "symd-link-to-dir"
-        $junctionToDir = Join-Path $TestPath "junction-to-dir"
+        $hardLinkToFile = Join-Path $TestPath "[hard-to-file].txt"
```
Since we haven't explicit tests please add new comment before the line that we intentionally use wildcard chars in follow names for testing.

GameMicrowave committed:
Update FileSystem.Tests.ps1

if file name contains wildcard character，the new-item should succeed on creation file link

GameMicrowave committed:
Merge branch 'master' into master

GameMicrowave committed:
Update FileSystem.Tests.ps1

Add comment

GameMicrowave committed:
Update FileSystem.Tests.ps1

GameMicrowave committed:
Update FileSystem.Tests.ps1

GameMicrowave committed:
Merge branch 'master' into master

iSazonov commented:
/azp run PowerShell-CI-linux-packaging, PowerShell-Windows-Packaging-CI

iSazonov commented:
@GameMicrowave Thanks for your contribution!
